pkgs/build-support/docker/stream_layered_image.py

```diff
@@ -354,7 +354,8 @@ def main():
         image_json = {
             "created": datetime.isoformat(created),
             "architecture": conf["architecture"],
-            "os": "linux",
+            "variant": conf["variant"],
+            "os": conf["os"],
```

Comment:
> Ultimately, any system should be capable of streaming any image, because there's nothing system-specific about the streaming script

I completely agree! That's why I think like `streamLayeredImage` itself shouldn't come from `pkgsCross`, its contents should. This PR embeds the `os` and `variant` from `pkgsCross` means that where `streamLayeredImage` runs should be the be the target.

Reply:
@utdemir That is a sufficient solution, but I think it would be nice to have a system-independent interface between
 - image contents and config
 - deployment tools

By making it system-independent, details about where the deployment tool runs don't seep into the image expressions.

Until we have that, we should document your solution.

Reply:
@utdemir Sorry for being late, I'm a little bit busy these times. I've updated the doc according to your request (I've not yet committed my changes), but there is one part I'm not sure to fully understand in your problem.

> So, when I am on a x86_64 machine, but want to produce Docker images that work on armv7, this is not what I want to use. Because if I run pkgsCross.armv7l-hf-multiplatform.dockerTools.streamLayeredImage, the streamLayeredImage executable that will emit the Docker image won't work on my computer

Despite what you claimed, I'm able to build and run the executable emitted by `streamLayeredImage` on my x86_64 computer, using `pkgs.pkgsCross.aarch64-multiplatform` to build and fetch the executable / docker image and then copy and successfully run the produced image on my PI4. Since the line you commented (down below) don't restrict the platforms it should be runs on (at least if I'm right). They are just being used to create the manifest config file for the image.

https://github.com/NixOS/nixpkgs/blob/84cf42812e82cc2d8879a45bfba9e62a1f74dc00/pkgs/build-support/docker/stream_layered_image.py#L354-L358

<details>
<summary>
Custom file used for testing
</summary>
<p>

```nix
{ pkgs, ... }:

pkgs.pkgsCross.aarch64-multiplatform.dockerTools.streamLayeredImage {
  name = "test";
  tag = "0.0.1";
  contents = [ pkgs.pkgsCross.aarch64-multiplatform.hello ];

  created = "now";

  fromImage = pkgs.pkgsCross.aarch64-multiplatform.dockerTools.pullImage {
    imageName = "docker.io/library/busybox";
    imageDigest = "sha256:ef320ff10026a50cf5f0213d35537ce0041ac1d96e9b7800bafd8bc9eff6c693";
    sha256 = "sha256-BjGm4uS98GhigVhtgEMJ4qMayqek/+cMSrCz3F2J4UQ=";
    finalImageName = "docker.io/library/busybox";
    finalImageTag = "stable-uclibc";
  };

  config = {
    WorkingDir = "/";

    Entrypoint = [ "/bin/hello" ];
  };
}
```

</p>
</details>

Reply:
>Despite what you claimed, I'm able to build and run the executable emitted by streamLayeredImage on my x86_64 computer, using pkgs.pkgsCross.aarch64-multiplatform to build and fetch the executable / docker image and then copy and successfully run the produced image on my PI4.

>  Since the line you commented (down below) don't restrict the platforms it should be runs on (at least if I'm right)

Perfect, if this works, I am happy!

My intuition (which is admittedly not great for cross-platform) stuff that the `streamLayeredImage` executable from `pkgsCross.someOtherPlatform` would only work on `someOtherPlatform`. If it works as you described, ignore my comment!